### PR TITLE
Vertex

### PR DESCRIPTION
**Pull request by pbutti**
This pull request is include the conversion from LCIO Vertex to ROOT Vertex classes, filling the hpstr objects and added some plotting. 
- Covariance matrix of the vertex is not rotated - TODO

### CODE REVIEW

File: processors/src/utilities.cxx
Comment (mholtrop) [inline processors/src/utilities.cxx:37]: Should we change code like this to use std::unique_ptr, to reduce the risk of memory leaks? 
That would require, e.g. vertices_.push_back(std::move(vtx));  

It would probably be cleaner if this code was one of the Vertex constructors, rather than being in a utilities collection. That would allow vector.emplace_back() which saves an object copy.

File: analysis/src/TrackHistos.cxx
Comment (mholtrop) [inline analysis/src/TrackHistos.cxx:6]: This code block mostly seems to have the indentation level change from 4 spaces to 2 spaces.
What spacing standard is maintained in hpstr?  (I prefer the 2 space, but stuck with what was established.)
(Changing the spacing here obfuscates any other changes that may have been made.)

File: event/include/Event.h
Comment (mholtrop) [inline event/include/Event.h:89]: Why save for later and not just document?

File: event/src/Event.cxx
Comment (mholtrop) [inline event/src/Event.cxx:13]: I agree with not creating the TTree inside the event. This _does_ change behavior of the Event class though.

File: processing/src/Process.cxx
Comment (mholtrop) [inline processing/src/Process.cxx:99]: indentation.

File: processors/src/RefittedTracksProcessor.cxx
Comment (mholtrop) [inline processors/src/RefittedTracksProcessor.cxx:76]: IF buildVertex creates std::unique_ptr, replace with vertices_.push_back(std::move(vtx))

IF not going with an std::move(), consider these two lines to be replace with vertices_.emplace_back() and instead of buildVertex, have Vertex object created with a proper constructor.

**Review by mholtrop — APPROVED** [review]
My comments are mostly for trying to learn about some of the choices made. I would like to know. 

The suggested change of buildVertex to a Vertex constructor would lead to cleaner code, but that is just my opinion and does not need to hold up the merge. Alternate of std::unique_ptr and std::move to speed up code and guard against leaks (elsewhere, I don't think current code leaks.)

File: processors/src/utilities.cxx
Comment (pbutti) [reply processors/src/utilities.cxx:37]: Hi Maurik, 

I think that using smart pointers all around hipster should be the way to go. Totally agree in updating that. There are several places now where bare pointers are used. I think the best is to create an issue dedicated to that task. I'll create one.

File: analysis/src/TrackHistos.cxx
Comment (pbutti) [reply analysis/src/TrackHistos.cxx:6]: Hi Maurik, 

I tend to prefer 2 as well. Cameron and Omar prefer 4...
The standard is not decided (even if Cameron says it's 4).

File: event/include/Event.h
Comment (pbutti) [reply event/include/Event.h:89]: Hi Maurik, 

I have bunch of //TODO here and there in the code. Perhaps I should decide to open an issue and fix all of them. I'll do it.

File: processing/src/Process.cxx
Comment (pbutti) [reply processing/src/Process.cxx:99]: Thanks! Will fix in next push

File: processors/src/RefittedTracksProcessor.cxx
Comment (pbutti) [reply processors/src/RefittedTracksProcessor.cxx:76]: Hi Maurik, 

In principle I'd like to have a set of utilities classes that take care of converting LCIO objects to ROOT objects that can be used by different processors. 
Basically I'd like to have the object conversion and ntuple building done in separate code. That's why I don't want to have constructors in the processor but internal to utilities.

Comment (mholtrop) [reply processors/src/RefittedTracksProcessor.cxx:76]: Hi PF,
Thank you for the answers!
I noticed that the EcalDataProcessor and SVTDataProcessor for instance, store the objects in TClonesArray* while the TrackingProcessor and RefittedTrackingProcessor store objects in vector<>.  I don't actually know for sure which is "better" in terms of performance or general usability, but it seems odd to have two different ways of doing this. Once we have a lot of root files written, this will be difficult to change, so perhaps we should think a bit more about the pros and cons of each and then decide what is best. Just saw the other pull request, where you state that you prefer vector<>. Is that for usability or for performance reasons?

As for utilities, vs everything in a processor, vs a constructor on the object (that is part of the event library, not processors), I would prefer the latter, unless there is a reason to keep all LCIO out of "event". That last reason is actually solid, but in that case, we should consider fixing the cmake files to make is possible to compile the "event" library without having LCIO install at all. I.e. if LCIO is not on your system, hpstr only compiles the libevent.so. That way the root files can be read and analyzed without needing LCIO, simplifying use.

**Pull request merged**
